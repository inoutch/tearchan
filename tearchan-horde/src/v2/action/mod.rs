use crate::action::manager::TimeMilliseconds;
use crate::v2::Tick;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tearchan_ecs::component::EntityId;

pub mod collection;
pub mod manager;

pub const VALID_SESSION_ID: ActionSessionId = 0;

pub type ActionSessionId = u64;

#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub enum ActionType {
    Start {
        tick: Tick,
    },
    Update {
        start: TimeMilliseconds,
        end: TimeMilliseconds,
    },
    End {
        tick: Tick,
    },
}

impl ActionType {
    pub fn tick(&self) -> Option<Tick> {
        match self {
            ActionType::Start { tick, .. } => Some(*tick),
            ActionType::Update { .. } => None,
            ActionType::End { tick, .. } => Some(*tick),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Action<T> {
    raw: T,
    entity_id: EntityId,
    ty: ActionType,
}

impl<T> Action<T> {
    pub fn new(raw: T, entity_id: EntityId, ty: ActionType) -> Self {
        Self { raw, entity_id, ty }
    }

    pub fn tick(&self) -> Option<Tick> {
        self.ty.tick()
    }
}

pub type ArcAction<T> = Action<Arc<T>>;

impl<T> Action<T> {
    pub fn raw(&self) -> &T {
        &self.raw
    }

    pub fn replace<U>(&self, raw: U) -> Action<U> {
        Action {
            raw,
            entity_id: self.entity_id,
            ty: self.ty,
        }
    }

    pub fn entity_id(&self) -> EntityId {
        self.entity_id
    }
}
