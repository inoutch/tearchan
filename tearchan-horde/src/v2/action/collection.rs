use crate::v2::action::manager::ActionSessionValidator;
use crate::v2::action::{ActionSessionId, ArcAction};
use crate::v2::Tick;
use std::any::{Any, TypeId};
use std::collections::hash_map::Iter;
use std::collections::HashMap;
use tearchan_ecs::component::EntityId;

pub struct ActionMeta {
    pub type_id: TypeId,
    pub entity_id: EntityId,
    pub session_id: ActionSessionId,
    pub tick: Tick,
}

pub struct ActionAnyVec {
    metas: Vec<ActionMeta>,
    actions: Box<dyn Any>,
}

impl ActionAnyVec {
    pub fn new<T: 'static>() -> Self {
        ActionAnyVec {
            metas: vec![],
            actions: Box::new(Vec::<ArcAction<T>>::new()),
        }
    }
}

impl ActionAnyVec {
    pub fn cast<T: 'static>(&self, validator: &ActionSessionValidator) -> Vec<&ArcAction<T>> {
        self.actions
            .downcast_ref::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .iter()
            .zip(self.metas.iter())
            .filter_map(|(action, meta)| {
                if validator.validate(meta) {
                    Some(action)
                } else {
                    None
                }
            })
            .collect()
    }

    pub fn cast_cloned<T: 'static>(&self, validator: &ActionSessionValidator) -> Vec<ArcAction<T>> {
        self.actions
            .downcast_ref::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .iter()
            .zip(self.metas.iter())
            .filter_map(|(action, meta)| {
                if validator.validate(meta) {
                    Some(action.clone())
                } else {
                    None
                }
            })
            .collect()
    }

    fn push<T: 'static>(&mut self, action: ArcAction<T>, session_id: ActionSessionId, tick: Tick) {
        let type_id = TypeId::of::<T>();
        self.metas.push(ActionMeta {
            type_id,
            entity_id: action.entity_id,
            session_id,
            tick,
        });
        self.actions
            .downcast_mut::<Vec<ArcAction<T>>>()
            .expect("Invalid type")
            .push(action);
    }
}

#[derive(Default)]
pub struct TypedActionAnyMap {
    map: HashMap<TypeId, ActionAnyVec>,
}

impl TypedActionAnyMap {
    pub fn push<T>(&mut self, action: ArcAction<T>, session_id: ActionSessionId)
    where
        T: 'static,
    {
        let type_id = TypeId::of::<T>();
        let tick = action.tick().expect("Must have tick");
        self.map
            .entry(type_id)
            .or_insert_with(ActionAnyVec::new::<T>)
            .push(action, session_id, tick);
    }

    pub fn get<T>(&self, validator: &ActionSessionValidator) -> Option<Vec<&ArcAction<T>>>
    where
        T: 'static,
    {
        let vec = self.map.get(&TypeId::of::<T>())?.cast(validator);
        if vec.is_empty() {
            return None;
        }
        Some(vec)
    }

    pub fn get_cloned<T>(&self, validator: &ActionSessionValidator) -> Option<Vec<ArcAction<T>>>
    where
        T: 'static,
    {
        let vec = self.map.get(&TypeId::of::<T>())?.cast_cloned(validator);
        if vec.is_empty() {
            return None;
        }
        Some(vec)
    }

    pub fn iter(&self) -> Iter<'_, TypeId, ActionAnyVec> {
        self.map.iter()
    }
}

#[derive(Default)]
pub struct AnyActionVec {
    vec: Vec<Box<dyn Any>>,
}

impl AnyActionVec {
    pub fn cast<T: 'static>(&self) -> Vec<&ArcAction<T>> {
        self.vec
            .iter()
            .filter_map(|item| item.downcast_ref())
            .collect()
    }

    pub fn cast_cloned<T: 'static>(&self) -> Vec<ArcAction<T>> {
        self.vec
            .iter()
            .filter_map(|item| item.downcast_ref::<ArcAction<T>>())
            .cloned()
            .collect::<Vec<_>>()
    }

    pub fn push<T: 'static>(&mut self, item: T) {
        self.vec.push(Box::new(item));
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.vec.is_empty()
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.vec.len()
    }

    #[inline]
    fn replace(&mut self, index: usize, item: Box<dyn Any>) {
        if let Some(x) = self.vec.get_mut(index) {
            *x = item
        }
    }

    #[inline]
    fn pop(&mut self) -> Option<Box<dyn Any>> {
        self.vec.pop()
    }

    #[inline]
    fn push_as_raw(&mut self, raw: Box<dyn Any>) {
        self.vec.push(raw);
    }
}

#[derive(Default)]
pub struct TypedAnyActionMapGroupedByEntityId {
    map: HashMap<TypeId, AnyActionVec>,
    indices: HashMap<EntityId, (TypeId, usize)>,
    entities: HashMap<(TypeId, usize), EntityId>,
}

impl TypedAnyActionMapGroupedByEntityId {
    pub fn get<T>(&self) -> Option<Vec<&ArcAction<T>>>
    where
        T: 'static,
    {
        Some(self.map.get(&TypeId::of::<T>())?.cast())
    }

    pub fn get_cloned<T>(&self) -> Option<Vec<ArcAction<T>>>
    where
        T: 'static,
    {
        Some(self.map.get(&TypeId::of::<T>())?.cast_cloned())
    }

    pub fn insert<T>(&mut self, entity_id: EntityId, action: ArcAction<T>)
    where
        T: 'static,
    {
        let type_id = TypeId::of::<T>();

        self.remove(entity_id);
        let vec = self
            .map
            .entry(type_id)
            .or_insert_with(AnyActionVec::default);
        let index = vec.len();
        vec.push(action);
        self.entities.insert((type_id, index), entity_id);
        self.indices.insert(entity_id, (type_id, index));
    }

    pub fn insert_with_type_id(
        &mut self,
        entity_id: EntityId,
        action: Box<dyn Any>,
        type_id: TypeId,
    ) {
        self.remove(entity_id);

        let collection = self
            .map
            .entry(type_id)
            .or_insert_with(AnyActionVec::default);
        let index = collection.len();
        collection.push_as_raw(action);
        self.entities.insert((type_id, index), entity_id);
        self.indices.insert(entity_id, (type_id, index));
    }

    pub fn remove(&mut self, entity_id: EntityId) {
        if let Some((type_id, index)) = self.indices.remove(&entity_id) {
            self.entities.remove(&(type_id, index));

            let collection = self.map.get_mut(&type_id).unwrap();
            let last = collection.len() - 1;
            let last_item = collection.pop().unwrap();
            if index != last {
                let last_entity_id = self.entities.remove(&(type_id, last)).unwrap();
                self.indices.remove(&last_entity_id);

                collection.replace(index, last_item);
                self.entities.insert((type_id, index), last_entity_id);
                self.indices.insert(last_entity_id, (type_id, index));
            } else if index == 0 {
                self.map.remove(&type_id);
            }
        }
    }

    pub fn iter(&self) -> Iter<'_, TypeId, AnyActionVec> {
        self.map.iter()
    }
}

#[cfg(test)]
mod test {
    use crate::v2::action::collection::TypedAnyActionMapGroupedByEntityId;
    use crate::v2::action::{Action, ActionType};
    use std::sync::Arc;

    #[derive(Clone, Debug)]
    struct MoveState;

    #[derive(Clone, Debug)]
    struct JumpState;

    #[test]
    fn test_typed_any_vec_grouped_by_entities() {
        let mut collection = TypedAnyActionMapGroupedByEntityId::default();

        collection.insert(
            1,
            Action {
                raw: Arc::new(MoveState),
                entity_id: 1,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );

        collection.insert(
            2,
            Action {
                raw: Arc::new(JumpState),
                entity_id: 2,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );
        collection.insert(
            2,
            Action {
                raw: Arc::new(JumpState),
                entity_id: 2,
                ty: ActionType::Start {
                    tick: 0,
                    start: 0,
                    end: 0,
                },
            },
        );

        insta::assert_debug_snapshot!(collection.get::<MoveState>());
        insta::assert_debug_snapshot!(collection.get::<JumpState>());

        collection.remove(2);

        insta::assert_debug_snapshot!(collection.get::<MoveState>()); // should have one
        insta::assert_debug_snapshot!(collection.get::<JumpState>()); // should be empty

        collection.remove(1);

        insta::assert_debug_snapshot!(collection.get::<MoveState>()); // should be empty

        assert_eq!(collection.entities.len(), 0);
        assert_eq!(collection.indices.len(), 0);
        assert_eq!(collection.map.len(), 0);
    }
}
