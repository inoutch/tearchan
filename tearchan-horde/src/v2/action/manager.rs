use crate::action::manager::TimeMilliseconds;
use crate::v2::action::collection::{
    ActionMeta, TypedActionAnyMap, TypedAnyActionVecGroupedByEntityId,
};
use crate::v2::action::{Action, ActionSessionId, ActionType, VALID_SESSION_ID};
use std::any::{Any, TypeId};
use std::collections::{BTreeMap, BTreeSet, HashMap, VecDeque};
use std::sync::Arc;
use tearchan_ecs::component::EntityId;

#[derive(Default)]
pub struct ActionSessionValidator<'a> {
    contexts: Option<&'a HashMap<EntityId, ActionContext>>,
}

impl<'a> ActionSessionValidator<'a> {
    pub fn validate(&self, action: &ActionMeta) -> bool {
        let contexts = match self.contexts {
            None => return true,
            Some(contexts) => contexts,
        };
        contexts
            .get(&action.entity_id)
            .map(|context| {
                action.session_id == context.session_id
                    || action.session_id == VALID_SESSION_ID
                    || action.tick <= context.session_expired_at
            })
            .unwrap_or(false)
    }
}
// x1/4 -> 60 / 4 = 15
// 1 tick is 66ms
const TICK_DURATION: u64 = 1000 / 15;
type Tick = u64;

enum Event {
    Started {
        type_id: TypeId,
        update_action: Box<dyn Any>,
        running_end_tick: Tick,
    },
    Ended,
    Canceled,
}

struct ActionContext {
    last_tick: Tick,             // The last time of stacking all actions
    running_end_tick: Tick,      // The end time of running action
    session_id: ActionSessionId, // Use for validation action
    session_expired_at: Tick,    // Invalid session is expired at this
}

#[derive(Default)]
struct ActionQueueItem {
    map: TypedActionAnyMap,
    events: VecDeque<(EntityId, Event)>,
    cancels: BTreeSet<EntityId>,
}

pub struct PullActionResult {
    pub map: TypedActionAnyMap,
    pub cancels: BTreeSet<EntityId>,
}

#[derive(Default)]
pub struct ActionManager {
    next_time: TimeMilliseconds,
    next_tick: Tick,
    current_tick: Tick,
    actions: BTreeMap<Tick, ActionQueueItem>,
    update_actions: TypedAnyActionVecGroupedByEntityId,
    contexts: HashMap<EntityId, ActionContext>,
    vacated_entities: BTreeSet<EntityId>,
}

impl ActionManager {
    pub fn attach(&mut self, entity_id: EntityId) {
        debug_assert!(!self.contexts.contains_key(&entity_id));

        self.contexts.insert(
            entity_id,
            ActionContext {
                last_tick: self.current_tick,
                running_end_tick: self.current_tick,
                session_id: 1,
                session_expired_at: self.current_tick,
            },
        );
    }

    pub fn detach(&mut self, entity_id: EntityId) {
        self.contexts.remove(&entity_id);
    }

    pub fn update(&mut self, delta: TimeMilliseconds) {
        debug_assert!(self.vacated_entities.is_empty());

        self.next_time += delta;
        self.next_tick = self.next_time / TICK_DURATION;
    }

    pub fn enqueue<T>(&mut self, entity_id: EntityId, raw: Arc<T>, duration: TimeMilliseconds)
    where
        T: 'static,
    {
        let type_id = TypeId::of::<T>();
        let context = self
            .contexts
            .get_mut(&entity_id)
            .unwrap_or_else(|| panic!("entity of {} is not attached", entity_id));
        debug_assert!(self.current_tick <= context.running_end_tick);
        debug_assert!(context.running_end_tick <= context.last_tick);

        let start_tick = context.last_tick;
        let end_tick = start_tick + duration / TICK_DURATION;
        {
            let item = self
                .actions
                .entry(start_tick)
                .or_insert_with(Default::default);
            item.map.push(
                Action {
                    raw: Arc::clone(&raw),
                    entity_id,
                    ty: ActionType::Start { tick: start_tick },
                },
                context.session_id,
            );
            item.events.push_back((
                entity_id,
                Event::Started {
                    type_id,
                    update_action: Box::new(Action {
                        raw: Arc::clone(&raw),
                        entity_id,
                        ty: ActionType::Update {
                            start: start_tick * TICK_DURATION,
                            end: end_tick * TICK_DURATION,
                        },
                    }),
                    running_end_tick: end_tick,
                },
            ));
        }
        {
            let item = self
                .actions
                .entry(end_tick)
                .or_insert_with(Default::default);
            item.map.push(
                Action {
                    raw,
                    entity_id,
                    ty: ActionType::End { tick: end_tick },
                },
                context.session_id,
            );
            item.events.push_back((entity_id, Event::Ended));
        }
        context.last_tick = end_tick;
    }

    pub fn pull_actions(&mut self) -> Option<PullActionResult> {
        let (tick, mut item) = self.actions.pop_first()?;
        if tick > self.next_tick {
            self.current_tick = self.next_tick;
            self.actions.insert(tick, item);
            return None;
        }

        self.current_tick = tick;

        while let Some((entity_id, event)) = item.events.pop_front() {
            let context = self.contexts.get_mut(&entity_id).unwrap();
            match event {
                Event::Started {
                    type_id,
                    update_action: action,
                    running_end_tick,
                } => {
                    self.update_actions
                        .insert_with_type_id(entity_id, action, type_id);
                    context.running_end_tick = running_end_tick;
                }
                Event::Ended => {
                    debug_assert!(self.current_tick <= context.last_tick);
                    if context.last_tick == self.current_tick {
                        self.vacated_entities.insert(entity_id);
                    }
                }
                Event::Canceled => {
                    self.vacated_entities.insert(entity_id);
                }
            }
        }

        Some(PullActionResult {
            map: item.map,
            cancels: item.cancels,
        })
    }

    pub fn pull_updates(&self) -> &TypedAnyActionVecGroupedByEntityId {
        &self.update_actions
    }

    pub fn pull_vacated_entities(&mut self) -> BTreeSet<EntityId> {
        std::mem::take(&mut self.vacated_entities)
    }

    pub fn cancel(&mut self, entity_id: EntityId, immediate: bool) {
        self.update_actions.remove(entity_id);

        let context = self.contexts.get_mut(&entity_id).unwrap();
        context.session_id += 1;
        let tick = if immediate {
            self.current_tick
        } else {
            context.running_end_tick
        };
        context.last_tick = tick;
        context.running_end_tick = tick;
        context.session_expired_at = tick;

        let item = self.actions.entry(tick).or_insert_with(Default::default);
        item.cancels.insert(entity_id);
        item.events.push_back((entity_id, Event::Canceled));
    }

    pub fn validator(&self) -> ActionSessionValidator {
        ActionSessionValidator {
            contexts: Some(&self.contexts),
        }
    }

    #[inline]
    pub fn current_tick(&self) -> Tick {
        self.current_tick
    }
}

#[cfg(test)]
#[macro_use]
mod test {
    use crate::define_actions;
    use crate::v2::action::manager::{
        ActionManager, ActionSessionValidator, PullActionResult, Tick,
    };
    use crate::v2::action::ArcAction;
    use serde::{Deserialize, Serialize};
    use std::sync::Arc;

    #[derive(Clone, Debug, Deserialize, Serialize)]
    struct MoveState;

    #[derive(Clone, Debug, Deserialize, Serialize)]
    struct JumpState;

    #[derive(Clone, Debug, Deserialize, Serialize)]
    struct TalkState;

    #[derive(Debug)]
    #[allow(dead_code)]
    struct SnapshotResult<'a> {
        tag: &'static str,
        current_tick: Tick,
        move_actions: Option<Vec<&'a ArcAction<MoveState>>>,
        jump_actions: Option<Vec<&'a ArcAction<JumpState>>>,
        talk_actions: Option<Vec<&'a ArcAction<TalkState>>>,
    }

    impl<'a> SnapshotResult<'a> {
        pub fn from_result(
            tag: &'static str,
            current_tick: Tick,
            result: &'a PullActionResult,
            validator: &ActionSessionValidator,
        ) -> Self {
            SnapshotResult {
                tag,
                current_tick,
                move_actions: result.map.get(validator),
                jump_actions: result.map.get(validator),
                talk_actions: result.map.get(validator),
            }
        }
    }

    define_actions!(
        TestAction,
        (Move, MoveState),
        (Jump, JumpState),
        (Talk, TalkState)
    );

    fn snapshot(manager: &mut ActionManager) {
        while let Some(actions) = manager.pull_actions() {
            insta::assert_debug_snapshot!(SnapshotResult {
                current_tick: manager.current_tick(),
                tag: "changes",
                move_actions: actions.map.get(&manager.validator()),
                jump_actions: actions.map.get(&manager.validator()),
                talk_actions: actions.map.get(&manager.validator())
            });
        }
        snapshot_update(manager);
    }

    fn snapshot_update(manager: &mut ActionManager) {
        let actions = manager.pull_updates();

        insta::assert_debug_snapshot!(SnapshotResult {
            current_tick: manager.current_tick(),
            tag: "updates",
            move_actions: actions.get(),
            jump_actions: actions.get(),
            talk_actions: actions.get()
        });
    }

    #[test]
    fn test() {
        // --- Millis: 0, Tick: 0 ---- //
        let mut manager = ActionManager::default();
        manager.attach(1);
        manager.attach(2);

        manager.enqueue(1, Arc::new(MoveState), 1000); // tick: 15
        manager.enqueue(1, Arc::new(JumpState), 2000); // tick: 30(45)
        manager.enqueue(1, Arc::new(TalkState), 1500); // tick: 22(67)

        manager.enqueue(2, Arc::new(MoveState), 500); // tick: 7
        manager.enqueue(2, Arc::new(JumpState), 1500); // tick: 22(29)
        manager.enqueue(2, Arc::new(TalkState), 2000); // tick: 30(59)

        snapshot(&mut manager);

        // --- Tick: 3 ---- //
        manager.update(200);

        snapshot(&mut manager);

        // --- Tick: 7 ---- //
        manager.update(300);

        snapshot(&mut manager);

        // --- Tick: 60
        manager.update(3328);

        snapshot(&mut manager);
    }

    #[test]
    fn test_vacated_entities() {
        let mut manager = ActionManager::default();
        manager.attach(1);
        manager.attach(2);

        manager.enqueue(1, Arc::new(MoveState), 1000); // tick: 15
        manager.enqueue(2, Arc::new(MoveState), 500); // tick: 7

        assert!(manager.pull_vacated_entities().is_empty());

        // Just ended time
        manager.update(500);

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        assert!(manager.pull_actions().is_none());
        assert_eq!(
            manager
                .pull_vacated_entities()
                .into_iter()
                .collect::<Vec<_>>(),
            vec![2]
        );

        manager.enqueue(2, Arc::new(JumpState), 2000);

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        assert!(manager.pull_actions().is_none());
        assert!(manager.pull_vacated_entities().is_empty());

        manager.update(10000);

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        assert_eq!(
            manager
                .pull_vacated_entities()
                .into_iter()
                .collect::<Vec<_>>(),
            vec![1]
        );

        manager.enqueue(1, Arc::new(TalkState), 3000);

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        assert!(manager.pull_vacated_entities().is_empty());

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));

        assert_eq!(
            manager
                .pull_vacated_entities()
                .into_iter()
                .collect::<Vec<_>>(),
            vec![2]
        );
    }

    #[test]
    fn test_cancel() {
        let mut manager = ActionManager::default();
        manager.attach(1);
        manager.attach(2);

        manager.enqueue(1, Arc::new(MoveState), 1000); // tick: 15
        manager.enqueue(1, Arc::new(JumpState), 2000); // tick: 30(45)
        manager.enqueue(1, Arc::new(TalkState), 1500); // tick: 22(67)

        manager.enqueue(2, Arc::new(MoveState), 500); // tick: 7
        manager.enqueue(2, Arc::new(JumpState), 1500); // tick: 22(29)
        manager.enqueue(2, Arc::new(TalkState), 2000); // tick: 30(59)
        manager.enqueue(2, Arc::new(MoveState), 3000); // tick: 45(104)

        manager.update(2112); // tick: 32

        let result = manager.pull_actions(); // tick: 0 - start1 start2
        assert_eq!(result.unwrap().cancels.len(), 0);
        let result = manager.pull_actions(); // tick: 7 - end2 start2
        assert_eq!(result.unwrap().cancels.len(), 0);
        let result = manager.pull_actions(); // tick: 15 - end1 end2
        assert_eq!(result.unwrap().cancels.len(), 0);
        let result = manager.pull_actions(); // tick: 29 - end2 start2
        assert_eq!(result.unwrap().cancels.len(), 0);

        assert_eq!(manager.current_tick(), 29);

        manager.cancel(1, true);

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));
        assert_eq!(result.cancels.into_iter().collect::<Vec<_>>(), vec![1]);
        assert_eq!(
            manager
                .pull_vacated_entities()
                .into_iter()
                .collect::<Vec<_>>(),
            vec![1]
        );

        manager.enqueue(1, Arc::new(TalkState), 2000); // tick: 30(59)

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));
        assert_eq!(result.cancels.len(), 0);

        assert!(manager.pull_actions().is_none());

        manager.update(396); // tick: 38

        assert!(manager.pull_actions().is_none());
        assert_eq!(manager.current_tick(), 38);

        manager.cancel(2, false);

        assert!(manager.pull_actions().is_none());

        manager.update(132); // tick: 40

        assert!(manager.pull_actions().is_none());
        assert_eq!(manager.current_tick(), 40);

        manager.update(1254); // tick: 59

        let result = manager.pull_actions().unwrap();
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));
        assert_eq!(result.cancels.len(), 0);
        assert_eq!(manager.pull_vacated_entities().len(), 0);

        let result = manager.pull_actions().unwrap();
        assert_eq!(manager.current_tick(), 59);
        insta::assert_debug_snapshot!(SnapshotResult::from_result(
            "changes",
            manager.current_tick(),
            &result,
            &manager.validator()
        ));
        assert_eq!(result.cancels.into_iter().collect::<Vec<_>>(), vec![2]);
        assert_eq!(
            manager
                .pull_vacated_entities()
                .into_iter()
                .collect::<Vec<_>>(),
            vec![1, 2]
        );
    }

    #[test]
    fn test_zero_time_action() {}

    #[test]
    fn test_detach() {}

    #[test]
    fn test_serialization() {}
}
