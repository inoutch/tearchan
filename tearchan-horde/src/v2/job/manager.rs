use crate::action::manager::TimeMilliseconds;
use crate::v2::action::collection::{TypedActionAnyMap, TypedAnyActionMapGroupedByEntityId};
use crate::v2::action::manager::{
    ActionManager, ActionManagerConverter, ActionManagerData, ActionManagerError,
    ActionSessionValidator,
};
use crate::v2::action::Action;
use crate::v2::job::HordeInterface;
use crate::v2::Tick;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, VecDeque};
use std::sync::Arc;
use tearchan_ecs::component::EntityId;
use tearchan_ecs::entity::manager::ENTITY_REMAPPER;

pub struct JobManager<T: HordeInterface> {
    action_manager: ActionManager,
    jobs: HashMap<EntityId, Vec<T::Job>>,
}

impl<T> Default for JobManager<T>
where
    T: HordeInterface,
{
    fn default() -> Self {
        JobManager {
            action_manager: Default::default(),
            jobs: Default::default(),
        }
    }
}

impl<T> JobManager<T>
where
    T: HordeInterface,
{
    pub fn run(&mut self, provider: &mut T, delta: TimeMilliseconds) {
        while !self.action_manager.get_vacated_entities().is_empty() {
            self.run_actions(provider);
        }
        self.action_manager.update(delta);
        self.run_actions(provider);
    }

    #[inline]
    pub fn attach(&mut self, entity_id: EntityId) {
        self.action_manager.attach(entity_id);
        self.jobs.insert(entity_id, Vec::new());
    }

    #[inline]
    pub fn detach(&mut self, entity_id: EntityId) {
        self.action_manager.detach(entity_id);
        self.jobs.remove(&entity_id);
    }

    #[inline]
    pub fn interrupt<U>(&mut self, entity_id: EntityId, raw: Arc<U>, duration: TimeMilliseconds)
    where
        U: 'static,
    {
        self.action_manager.interrupt(entity_id, raw, duration);
    }

    pub fn to_data<U>(
        &self,
        converter0: fn(&TypedActionAnyMap, &ActionSessionValidator) -> Vec<Action<U>>,
        converter1: fn(&TypedAnyActionMapGroupedByEntityId) -> Vec<Action<U>>,
    ) -> JobManagerData<U, T::Job> {
        JobManagerData {
            action_manager_data: self.action_manager.to_data(converter0, converter1),
            jobs: self.jobs.clone(),
        }
    }

    pub fn load_data<U>(
        &mut self,
        mut data: JobManagerData<U, T::Job>,
        converter: fn(action: Action<U>, manager: &mut ActionManagerConverter),
    ) {
        self.action_manager
            .load_data(data.action_manager_data, converter);
        for (entity_id, job) in data.jobs.drain() {
            self.jobs.insert(ENTITY_REMAPPER.remap(entity_id), job);
        }
    }

    pub fn from_data<U>(
        data: JobManagerData<U, T::Job>,
        converter: fn(action: Action<U>, manager: &mut ActionManagerConverter),
    ) -> Result<Self, JobManagerError> {
        let action_manager = ActionManager::from_data(data.action_manager_data, converter)
            .map_err(JobManagerError::ActionManagerError)?;
        for (entity_id, _) in data.jobs.iter() {
            if !action_manager.has_some_actions(*entity_id) {
                return Err(JobManagerError::NotFoundEntity(*entity_id));
            }
        }
        Ok(JobManager {
            action_manager,
            jobs: data.jobs,
        })
    }

    #[inline]
    pub fn current_tick(&self) -> Tick {
        self.action_manager.current_tick()
    }

    fn run_actions(&mut self, provider: &mut T) {
        // Loop for each tick
        loop {
            let result_or_none = self.action_manager.pull_actions();
            if let Some(result) = &result_or_none {
                provider.on_change_tick(&result.map, &self.action_manager.validator());

                for entity_id in result.cancels.iter() {
                    provider.on_cancel_job(
                        *entity_id,
                        std::mem::take(self.jobs.get_mut(entity_id).unwrap()),
                    );
                }
            }

            let vacated_entities = self.action_manager.pull_vacated_entities();
            if vacated_entities.is_empty() && result_or_none.is_none() {
                break;
            }
            for entity_id in vacated_entities {
                let mut priority = 0;
                let mut job_queue: VecDeque<T::Job> = VecDeque::new();
                self.jobs.get_mut(&entity_id).unwrap().clear();
                job_queue.push_front(provider.on_first(entity_id, priority));

                while let Some(job) = job_queue.pop_front() {
                    self.jobs.get_mut(&entity_id).unwrap().push(job.clone());

                    let result =
                        provider.on_next(entity_id, job, &mut self.action_manager.controller());
                    if job_queue.is_empty() && !self.action_manager.has_some_actions(entity_id) {
                        // If the jobs and actions cannot be generated from the current job tree,
                        // change the priority and recreate the first job
                        priority += 1;
                        self.jobs.get_mut(&entity_id).unwrap().clear();
                        job_queue.push_front(provider.on_first(entity_id, priority));
                        continue;
                    }

                    if let Some(result) = result {
                        job_queue.push_back(result);
                    }
                }
            }
        }

        provider.on_change_time(
            self.action_manager.pull_updates(),
            self.action_manager.next_time(),
        );
    }
}

pub struct JobController<'a> {
    action_manager: &'a mut ActionManager,
}

impl<'a> JobController<'a> {
    #[inline]
    pub fn enqueue<T>(&mut self, entity_id: EntityId, raw: Arc<T>, duration: TimeMilliseconds)
    where
        T: 'static,
    {
        self.action_manager.enqueue(entity_id, raw, duration);
    }
}

#[derive(Serialize, Deserialize)]
pub struct JobManagerData<T, U> {
    action_manager_data: ActionManagerData<T>,
    jobs: HashMap<EntityId, Vec<U>>,
}

#[derive(Debug)]
pub enum JobManagerError {
    ActionManagerError(ActionManagerError),
    NotFoundEntity(EntityId),
}
