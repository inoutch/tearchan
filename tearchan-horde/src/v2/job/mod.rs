use crate::action::manager::TimeMilliseconds;
use crate::v2::action::collection::{TypedAnyActionMap, TypedAnyActionMapGroupedByEntityId};
use crate::v2::action::manager::ActionController;
use crate::v2::job::manager::JobController;
use tearchan_ecs::component::EntityId;

pub mod manager;

pub trait HordeInterface {
    type Job: Clone;

    fn on_change_tick(&mut self, map: &TypedAnyActionMap, controller: JobController<Self::Job>);

    fn on_change_time(&mut self, map: &TypedAnyActionMapGroupedByEntityId, time: TimeMilliseconds);

    fn on_cancel_job(&mut self, entity_id: EntityId, jobs: Vec<Self::Job>);

    fn on_first(&self, entity_id: EntityId, priority: u32) -> Self::Job;

    fn on_next(
        &self,
        entity_id: EntityId,
        job: Self::Job,
        controller: &mut ActionController,
    ) -> Option<Self::Job>;
}
