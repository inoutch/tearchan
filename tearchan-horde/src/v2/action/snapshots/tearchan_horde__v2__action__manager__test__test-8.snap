---
source: tearchan-horde/src/v2/action/manager.rs
expression: "SnapshotResult {\n    current_tick: manager.current_tick(),\n    tag: \"changes\",\n    move_actions: actions.map.get(&manager.validator()),\n    jump_actions: actions.map.get(&manager.validator()),\n    talk_actions: actions.map.get(&manager.validator()),\n}"

---
SnapshotResult {
    tag: "changes",
    current_tick: 45,
    move_actions: None,
    jump_actions: Some(
        [
            Action {
                raw: JumpState,
                entity_id: 1,
                ty: End {
                    start: 15,
                    end: 45,
                },
            },
        ],
    ),
    talk_actions: Some(
        [
            Action {
                raw: TalkState,
                entity_id: 1,
                ty: Start {
                    start: 45,
                    end: 67,
                },
            },
        ],
    ),
}
